### PR TITLE
Setup CRUD Scaffold

### LINKED ISSUE #11: Creating articles CRUD
- [ ] /create articles
- [ ] [/read articles](https://github.com/girijaasoni/newsletterApplicaton/issues/2)
- [ ] /update articles
- [ ] /delete articles

### PR DESCRIPTION
**Pull request by archanaserver**
This PR contains the following changes:

- A new `aricles` view which perform all the crud operations. You can view this folder from `/app/views/articles`
- Also added the routes `resources :articles` in the `routes.rb` file.
- Go to `localhost:3000/articles` and you will see the list of the articles, also you can create new articles.
- Here are the screenshots from my local:

Screenshot from 2022-12-06 13-53-08
Screenshot from 2022-12-06 13-53-17

Resolves #11 #2 Setup and installed CRUD scaffold into the project.

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ require:
     - rubocop-rails
     - rubocop-rake
     - rubocop-minitest
+    - rubocop-i18n
```
Comment (Ron-Lavi) [inline .rubocop.yml]: are we really using any i18n features in this project?

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
+class ArticlesController < ApplicationController
+  before_action :set_article, only: %i[show edit update destroy]
```
Comment (Ron-Lavi) [inline app/controllers/articles_controller.rb]: maybe we should use `except` here instead of `only`

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        format.json { render json: @article.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH/PUT /articles/1 or /articles/1.json
+  def update
+    respond_to do |format|
+      if @article.update(article_params)
+        format.html { redirect_to article_url(@article), notice: _("Article was successfully updated.") }
```
Comment (Ron-Lavi) [inline app/controllers/articles_controller.rb]: All of the HTML responses are OK for now,
though as a follow-up we will re-write it in React

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  def edit
+  end
+
+  # POST /articles or /articles.json
+  def create
+    @article = Article.new(article_params)
+
+    respond_to do |format|
+      if @article.save
+        format.html { redirect_to article_url(@article), notice: _("Article was successfully created.") }
```
Comment (Ron-Lavi) [inline app/controllers/articles_controller.rb]: I get an error: `undefined method `_' for #<ArticlesController:0x0000000000bf40>`
do we need this i18n feature?

Comment (Ron-Lavi) [reply app/controllers/articles_controller.rb]: for easier debugging, can you also add https://github.com/pry/pry to the project?
opened an issue in https://github.com/girijaasoni/newsletterApplicaton/issues/22

File: config/routes.rb [hunk 5]
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
+  resources :articles
```
Comment (Ron-Lavi) [inline config/routes.rb]: can it move below the root route?

File: db/migrate/20221206080208_create_articles.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class CreateArticles < ActiveRecord::Migration[7.0]
+  def change
+    create_table :articles do |t|
+      t.string :creator_name
+      t.string :article_title
+      t.text :article_body
```
Comment (Ron-Lavi) [inline db/migrate/20221206080208_create_articles.rb]: maybe we can simplify it a bit, e.g:
```suggestion
      t.string :creator
      t.string :title
      t.text :body
```

File: test/controllers/articles_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  test "should destroy article" do
+    assert_difference("Article.count", -1) do
+      delete article_url(@article)
+    end
+
+    assert_redirected_to articles_url
+  end
+end
```
Comment (Ron-Lavi) [inline test/controllers/articles_controller_test.rb:57]: Since we added the test, can we also add a CI for those? we can re-use the `rubocop.yml` and rename it to `ruby_ci.yml`

**Review by Ron-Lavi — COMMENTED** [review]
Great work @archanaserver ! left a few inline comments

File: app/controllers/articles_controller.rb (hunk 4 above)
Comment (archanaserver) [reply app/controllers/articles_controller.rb]: I have removed the i18n feature from the rubocop rule and i think we don't need it and it's working fine.

File: .github/workflows/ruby-ci.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+name: Ruby-CI
+on: [pull_request]
+permissions:
+  contents: read
+
+jobs:
+  test:
```
Comment (Ron-Lavi) [inline .github/workflows/ruby-ci.yml]: You will need to add the service of PostgreSQL here as well, e.g: https://github.com/theforeman/foreman_ansible/blob/master/.github/workflows/ruby_tests.yml#L24

and after that you need to run the DB preparation commands:
https://github.com/theforeman/foreman_ansible/blob/master/.github/workflows/ruby_tests.yml#L66-L68

if this is too much it can be on another PR.

File: .github/workflows/ruby-ci.yml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Setup Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 3.1.2
+          bundler-cache: true
+      - name: Run tests
+        run: bundle exec rake
```
Comment (Ron-Lavi) [inline .github/workflows/ruby-ci.yml]: the command should be `bundle exec rake test` :)

Comment (archanaserver) [reply .github/workflows/ruby-ci.yml]: Oh yes, Thank you :)

File: .github/workflows/ruby-ci.yml (hunk 8 above)
Comment (archanaserver) [reply .github/workflows/ruby-ci.yml]: Created this new issue https://github.com/girijaasoni/newsletterApplicaton/issues/23 for the same. Will update the CI test workflow there.

File: db/schema.rb [hunk 10]
```diff
@@ -10,8 +10,16 @@
...
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.0].define(version: 0) do
+ActiveRecord::Schema[7.0].define(version: 2022_12_06_080208) do
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
+  create_table "articles", force: :cascade do |t|
+    t.string "creator_name"
+    t.string "article_title"
```
Comment (Ron-Lavi) [inline db/schema.rb]: Can you update the schema?

File: db/seeds.rb [hunk 11]
```diff
@@ -5,3 +5,23 @@
 #
 #   movies = Movie.create([{ name: "Star Wars" }, { name: "Lord of the Rings" }])
 #   Character.create(name: "Luke", movie: movies.first)
+
+Article.destroy_all
+
+Article.create!([{
+                  creator_name: "John Watson",
+                  article_title: "Ruby on Rails with PostgreSQL",
+                  article_body: "This is the main body of the article."
```
Comment (Ron-Lavi) [inline db/seeds.rb]: Can you update the seeds with the correct table fields?

File: test/fixtures/articles.yml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  creator_name: MyString
+  article_title: MyString
+  article_body: MyText
```
Comment (Ron-Lavi) [inline test/fixtures/articles.yml]: also needs an update :)

**Comment by archanaserver** [comment]
@Ron-Lavi it's updated now, also all the checks passed.

File: app/views/articles/_article.html.erb [hunk 13]
```diff
@@ -0,0 +1,17 @@
+<div id="<%= dom_id article %>">
+  <p>
+    <strong>Creator name:</strong>
+    <%= article.creator_name %>
```
Comment (Ron-Lavi) [inline app/views/articles/_article.html.erb]: can you also fix all of the changes of the migration in the templates?

**Review by Ron-Lavi — APPROVED** [review]
thanks @archanaserver ! LGTM :)

**Comment by Ron-Lavi** [comment]
#26 got merged, can you rebase this PR?
